xmlrpc y2038 problem

(originally filed in the openSUSE Bugzilla](https://bugzilla.suse.com/show_bug.cgi?id=1203311)):

While working on reproducible builds for openSUSE, I found that our python-softlayer package failed tests in 2038 with

```
 self = <xmlrpc.client.Marshaller object at 0x7fa5770e91f0>, value = 2167459529
 write = <built-in method append of list object at 0x7fa577194740>
 
     def dump_long(self, value, write):
         if value > MAXINT or value < MININT:
 >           raise OverflowError("int exceeds XML-RPC limits")
 E           OverflowError: int exceeds XML-RPC limits
 
 /usr/lib64/python3.8/xmlrpc/client.py:539: OverflowError
 =========================== short test summary info ============================
 FAILED tests/CLI/modules/cdn_tests.py::CdnTests::test_detail_account - Overfl...
 FAILED tests/managers/cdn_tests.py::CDNTests::test_detail_usage_metric - Over...
 =========== 2 failed, 1831 passed, 6 skipped, 5 deselected in 38.01s ===========
```

To Reproduce:
```
osc co openSUSE:Factory/python-softlayer && cd $_
osc build --noservice --trust-all-projects --vm-type=kvm \
  --alternative-project=home:bmwiedemann:reproducible \
  --build-opt=--vm-custom-opt="-rtc base=2038-01-20T00:00:00" \
  openSUSE_Tumbleweed x86_64
```

(basically run with the machine hardware clock set to post-2038)

**Expected behavior**
Test suite should pass.


**Version**
6.1.2 from tarball.

The current #1765 does not help as it fixes nothing that would only break in 2038.

Here is a full backtrace:
```
 /usr/lib64/python3.8/xmlrpc/client.py:539: OverflowError
 ----------------------------- Captured stdout call -----------------------------
 int exceeds XML-RPC limits
 ______________________ CDNTests.test_detail_usage_metric _______________________

 self = <tests.managers.cdn_tests.CDNTests testMethod=test_detail_usage_metric>

     def test_detail_usage_metric(self):
 >       self.cdn_client.get_usage_metrics(12345, history=30, frequency="aggregate")

 tests/managers/cdn_tests.py:32:
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 SoftLayer/managers/cdn.py:163: in get_usage_metrics
     usage = self.cdn_metrics.getMappingUsageMetrics(unique_id, self._start_date, self._end_date, frequency)
 SoftLayer/API.py:610: in call_handler
     return self(name, *args, **kwargs)
 SoftLayer/API.py:578: in call
     return self.client.call(self.name, name, *args, **kwargs)
 SoftLayer/API.py:302: in call
     return self.transport(request)
 SoftLayer/transports/timing.py:22: in __call__
     result = self.transport(call)
 SoftLayer/transports/xmlrpc.py:81: in __call__
     request.payload = xmlrpc.client.dumps(tuple(largs),
 /usr/lib64/python3.8/xmlrpc/client.py:968: in dumps
     data = m.dumps(params)
 /usr/lib64/python3.8/xmlrpc/client.py:501: in dumps
     dump(v, write)
 /usr/lib64/python3.8/xmlrpc/client.py:523: in __dump
     f(self, value, write)
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

 self = <xmlrpc.client.Marshaller object at 0x7f696fc6c520>, value = 2170227406
 write = <built-in method append of list object at 0x7f697658da00>
 
     def dump_long(self, value, write):
         if value > MAXINT or value < MININT:
 >           raise OverflowError("int exceeds XML-RPC limits")
 E           OverflowError: int exceeds XML-RPC limits

 /usr/lib64/python3.8/xmlrpc/client.py:539: OverflowError
```

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -48,5 +48,5 @@ jobs:
       with:
         user: __token__
         password: ${{ secrets.CGALLO_PYPI }}
-        repository_url: https://pypi.org/legacy/
+        repository_url: https://upload.pypi.org/legacy/
 

--- a/SoftLayer/CLI/cdn/detail.py
+++ b/SoftLayer/CLI/cdn/detail.py
@@ -41,6 +41,6 @@ def cli(env, unique_id, history):
     table.add_row(['status', cdn_mapping['status']])
     table.add_row(['total_bandwidth', total_bandwidth])
     table.add_row(['total_hits', total_hits])
-    table.add_row(['hit_radio', hit_ratio])
+    table.add_row(['hit_ratio', hit_ratio])
 
     env.fout(table)

--- a/SoftLayer/fixtures/SoftLayer_Network_CdnMarketplace_Configuration_Mapping.py
+++ b/SoftLayer/fixtures/SoftLayer_Network_CdnMarketplace_Configuration_Mapping.py
@@ -11,7 +11,7 @@ listDomainMappings = [
         "path": "/",
         "protocol": "HTTP",
         "status": "CNAME_CONFIGURATION",
-        "uniqueId": "9934111111111",
+        "uniqueId": "11223344",
         "vendorName": "akamai"
     }
 ]
@@ -28,7 +28,7 @@ listDomainMappingByUniqueId = [
         "path": "/",
         "protocol": "HTTP",
         "status": "CNAME_CONFIGURATION",
-        "uniqueId": "9934111111111",
+        "uniqueId": "11223344",
         "vendorName": "akamai"
     }
 ]
@@ -41,7 +41,7 @@ updateDomainMapping = [
         "performanceConfiguration": "Large file optimization",
         "protocol": "HTTP",
         "respectHeaders": True,
-        "uniqueId": "424406419091111",
+        "uniqueId": "11223344",
         "vendorName": "akamai",
         "header": "www.test.com",
         "httpPort": 83,

--- a/SoftLayer/fixtures/SoftLayer_Network_CdnMarketplace_Metrics.py
+++ b/SoftLayer/fixtures/SoftLayer_Network_CdnMarketplace_Metrics.py
@@ -6,9 +6,9 @@ getMappingUsageMetrics = [
             "HitRatio"
         ],
         "totals": [
-            "0.0",
-            "0",
-            "0.0"
+            1.0,
+            3,
+            2.0
         ],
         "type": "TOTALS"
     }

--- a/tests/CLI/modules/cdn_tests.py
+++ b/tests/CLI/modules/cdn_tests.py
@@ -21,7 +21,7 @@ class CdnTests(testing.TestCase):
                            'domain': 'test.example.com',
                            'origin': '1.1.1.1',
                            'status': 'CNAME_CONFIGURATION',
-                           'unique_id': '9934111111111',
+                           'unique_id': '11223344',
                            'vendor': 'akamai'}]
                          )
 
@@ -29,19 +29,12 @@ class CdnTests(testing.TestCase):
         result = self.run_command(['cdn', 'detail', '--history=30', '1245'])
 
         self.assert_no_fail(result)
-        self.assertEqual(json.loads(result.output),
-                         {'hit_radio': '0.0 %',
-                          'hostname': 'test.example.com',
-                          'origin': '1.1.1.1',
-                          'origin_type': 'HOST_SERVER',
-                          'path': '/',
-                          'protocol': 'HTTP',
-                          'provider': 'akamai',
-                          'status': 'CNAME_CONFIGURATION',
-                          'total_bandwidth': '0.0 GB',
-                          'total_hits': '0',
-                          'unique_id': '9934111111111'}
-                         )
+        api_results = json.loads(result.output)
+        self.assertEqual(api_results['hit_ratio'], '2.0 %')
+        self.assertEqual(api_results['total_bandwidth'], '1.0 GB')
+        self.assertEqual(api_results['total_hits'], 3)
+        self.assertEqual(api_results['hostname'], 'test.example.com')
+        self.assertEqual(api_results['protocol'], 'HTTP')
 
     def test_purge_content(self):
         result = self.run_command(['cdn', 'purge', '1234',
@@ -122,7 +115,7 @@ class CdnTests(testing.TestCase):
         self.assertEqual('include: test', header_result['Cache key optimization'])
 
     def test_edit_cache_by_uniqueId(self):
-        result = self.run_command(['cdn', 'edit', '9934111111111', '--cache', 'include-specified', '--cache', 'test'])
+        result = self.run_command(['cdn', 'edit', '11223344', '--cache', 'include-specified', '--cache', 'test'])
         self.assert_no_fail(result)
         header_result = json.loads(result.output)
         self.assertEqual('include: test', header_result['Cache key optimization'])

--- a/tests/CLI/modules/cdn_tests.py
+++ b/tests/CLI/modules/cdn_tests.py
@@ -4,7 +4,9 @@
 
     :license: MIT, see LICENSE for more details.
 """
+import datetime
 import json
+from unittest import mock as mock
 
 from SoftLayer.CLI import exceptions
 from SoftLayer import testing
@@ -25,7 +27,9 @@ class CdnTests(testing.TestCase):
                            'vendor': 'akamai'}]
                          )
 
-    def test_detail_account(self):
+    @mock.patch('SoftLayer.utils.days_to_datetime')
+    def test_detail_account(self, mock_now):
+        mock_now.return_value = datetime.datetime(2020, 1, 1)
         result = self.run_command(['cdn', 'detail', '--history=30', '1245'])
 
         self.assert_no_fail(result)

--- a/tests/managers/cdn_tests.py
+++ b/tests/managers/cdn_tests.py
@@ -4,12 +4,12 @@
 
     :license: MIT, see LICENSE for more details.
 """
+import datetime
+from unittest import mock as mock
 
 from SoftLayer import fixtures
 from SoftLayer.managers import cdn
 from SoftLayer import testing
-from unittest import mock as mock
-import datetime
 
 
 class CDNTests(testing.TestCase):
@@ -51,7 +51,6 @@ class CDNTests(testing.TestCase):
             OverflowError,
             self.cdn_client.get_usage_metrics, 12345, history=30, frequency="aggregate"
         )
-        
 
     def test_get_origins(self):
         self.cdn_client.get_origins("12345")

--- a/tests/managers/cdn_tests.py
+++ b/tests/managers/cdn_tests.py
@@ -8,6 +8,8 @@
 from SoftLayer import fixtures
 from SoftLayer.managers import cdn
 from SoftLayer import testing
+from unittest import mock as mock
+import datetime
 
 
 class CDNTests(testing.TestCase):
@@ -28,7 +30,9 @@ class CDNTests(testing.TestCase):
                                 'listDomainMappingByUniqueId',
                                 args=args)
 
-    def test_detail_usage_metric(self):
+    @mock.patch('SoftLayer.utils.days_to_datetime')
+    def test_detail_usage_metric(self, mock_now):
+        mock_now.return_value = datetime.datetime(2020, 1, 1)
         self.cdn_client.get_usage_metrics(12345, history=30, frequency="aggregate")
 
         args = (12345,
@@ -38,6 +42,16 @@ class CDNTests(testing.TestCase):
         self.assert_called_with('SoftLayer_Network_CdnMarketplace_Metrics',
                                 'getMappingUsageMetrics',
                                 args=args)
+
+    # Does this still work in 2038 ? https://github.com/softlayer/softlayer-python/issues/1764 for context
+    @mock.patch('SoftLayer.utils.days_to_datetime')
+    def test_detail_usage_metric_future(self, mock_now):
+        mock_now.return_value = datetime.datetime(2040, 1, 1)
+        self.assertRaises(
+            OverflowError,
+            self.cdn_client.get_usage_metrics, 12345, history=30, frequency="aggregate"
+        )
+        
 
     def test_get_origins(self):
         self.cdn_client.get_origins("12345")
@@ -105,7 +119,7 @@ class CDNTests(testing.TestCase):
                                 args=args)
 
     def test_cdn_edit(self):
-        identifier = '9934111111111'
+        identifier = '11223344'
         header = 'www.test.com'
         result = self.cdn_client.edit(identifier, header=header)
 
@@ -116,7 +130,7 @@ class CDNTests(testing.TestCase):
             'SoftLayer_Network_CdnMarketplace_Configuration_Mapping',
             'updateDomainMapping',
             args=({
-                'uniqueId': '9934111111111',
+                'uniqueId': '11223344',
                 'originType': 'HOST_SERVER',
                 'protocol': 'HTTP',
                 'path': '/',

